src/Orleans.Core/Runtime/CallbackData.cs

```diff
@@ -38,61 +77,100 @@ public bool IsExpired(long currentTimestamp)
 
         public void OnTimeout(TimeSpan timeout)
         {
+            // set/clear activityid
             if (this.IsCompleted)
                 return;
-            var msg = this.Message; // Local working copy
-
-            string messageHistory = msg.GetTargetHistory();
-            string errorMsg = $"Response did not arrive on time in {timeout} for message: {msg}. Target History is: {messageHistory}.";
-            this.shared.Logger.Warn(ErrorCode.Runtime_Error_100157, "{0} About to break its promise.", errorMsg);
+            var previousActivityId = EventSource.CurrentThreadActivityId;
+            try
+            {
+                EventSource.SetCurrentThreadActivityId(this.Message.ActivityId);
```

Comment:
Note that you can set + get the activity id in one call: https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventsource.setcurrentthreadactivityid?view=netframework-4.8#System_Diagnostics_Tracing_EventSource_SetCurrentThreadActivityId_System_Guid_System_Guid__

Reply:
yeah I know. but it won't work with try-finally.